pkg/segment/writer/segmetarw.go

```diff
@@ -0,0 +1,552 @@
+	"errors"
+	"fmt"
+	"os"
+	"path"
+	"sync"
+
+	"github.com/siglens/siglens/pkg/config"
+	"github.com/siglens/siglens/pkg/hooks"
+	pqsmeta "github.com/siglens/siglens/pkg/segment/query/pqs/meta"
+	"github.com/siglens/siglens/pkg/segment/structs"
+	"github.com/siglens/siglens/pkg/utils"
+	log "github.com/sirupsen/logrus"
+)
+
+const ONE_MiB = 1024 * 1024
+
+var smrLock sync.RWMutex = sync.RWMutex{}
+var localSegmetaFname string
+
+var SegmetaFilename = "segmeta.json"
+
+func initSmr() {
+
+	localSegmetaFname = GetLocalSegmetaFName()
+
+	fd, err := os.OpenFile(localSegmetaFname, os.O_RDONLY, 0666)
+	if err != nil {
+		if errors.Is(err, os.ErrNotExist) {
+			// for first time during bootup this will occur
+			_, err := os.OpenFile(localSegmetaFname, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0644)
+			if err != nil {
+				log.Errorf("initSmr: failed to open a new filename=%v: err=%v", localSegmetaFname, err)
+				return
+			}
+		}
+		return
+	}
+	fd.Close()
+}
+
+func getSegFullMetaFnameFromSegkey(segkey string) string {
+	return fmt.Sprintf("%s.sfm", segkey)
+}
+
+func ReadSegmeta(smFilename string) []*structs.SegMeta {
+	smrLock.RLock()
+	retVal, err := getAllSegmetas(smFilename)
+	smrLock.RUnlock()
+	if err != nil {
+		log.Errorf("ReadSegmeta: getallsegmetas err=%v ", err)
+	}
+	return retVal
+}
+
+// read only the current nodes segmeta
+func ReadLocalSegmeta(readFullMeta bool) []*structs.SegMeta {
+
+	smrLock.RLock()
+	retVal, err := getAllSegmetas(localSegmetaFname)
+	smrLock.RUnlock()
+	if err != nil {
+		log.Errorf("ReadLocalSegmeta: getallsegmetas err=%v ", err)
+		return retVal
+	}
+
+	if !readFullMeta {
+		return retVal
+	}
+
+	// continue reading/merging from individual segfiles
+	for _, smentry := range retVal {
+		sfmData, _ := readSfm(smentry.SegmentKey)
+		if sfmData == nil {
+			continue
+		}
+		if smentry.AllPQIDs == nil {
+			smentry.AllPQIDs = sfmData.AllPQIDs
+		} else {
+			utils.MergeMapsRetainingFirst(smentry.AllPQIDs, sfmData.AllPQIDs)
+		}
+		if smentry.ColumnNames == nil {
+			smentry.ColumnNames = sfmData.ColumnNames
+		} else {
+			utils.MergeMapsRetainingFirst(smentry.ColumnNames, sfmData.ColumnNames)
+		}
+	}
+	return retVal
+}
+
+func readSfm(segkey string) (*structs.SegFullMeta, error) {
+
+	sfmFname := getSegFullMetaFnameFromSegkey(segkey)
+
+	sfmBytes, err := os.ReadFile(sfmFname)
+	if err != nil {
+		if !os.IsNotExist(err) {
+			log.Errorf("readSfm: Cannot read sfm File: %v, err: %v", sfmFname, err)
+		}
+		return nil, err
+	}
+	sfm := &structs.SegFullMeta{}
+	if err := json.Unmarshal(sfmBytes, sfm); err != nil {
+		log.Errorf("readSfm: Error unmarshalling sfm file: %v, data: %v err: %v",
+			sfmFname, string(sfmBytes), err)
+		return nil, err
+	}
+	return sfm, nil
+}
+
+func writeSfm(segkey string, sfmData *structs.SegFullMeta) {
+
+	// create a separate individual file for SegFullMeta
+	sfmFname := getSegFullMetaFnameFromSegkey(segkey)
+	sfmFd, err := os.OpenFile(sfmFname, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0644)
```

Comment:
Do you need to get a write lock for this file? Is a lock not needed because this only happens on rotation and rotation only happens once? Is `readSfm` guaranteed to only happen after `writeSfm` (if it happens concurrently, `readSfm` might get partial JSON which it can't decode)

Reply:
`writeSfm` only happens once when we rotate the segment, until then the readSfm cannot be called because this segment has not been rotated yet (and it is not present in segmeta.json) so the reader/populateMicroIndices code will not see this seg hence they can't call the readSfm func